[MNT] Purging similarity search from distance as argument numba functions

baraline opened the pull request:
#### Reference Issues/PRs

Fixes #2071, #2163, #2250
#### What does this implement/fix? Explain your changes.

- Removes the naive distance and matrix profiles functions that were using a numba function (distance) as argument and some associated components in utils.
- Move some functions from RDST to general numba utils as I reuse them in similarity search to define a "naive" distance Euclidean profile to replace the old one, which was used in testing 
- Add tests for squared distance profile
- Uncomment tests that were commented because of #2071, and add some test exclusion due to issue with the estimator checking function giving wrong input shape for predict function of similarity search (3D fit, 2D predict).
- Remove some unused functions in numba utils
- Remove the fact that distance profiles were returned per channel, and then summed in `QuerySearch` estimator, now they are computed directly in the distance profile functions.
- Update similarity search image on docs
- Change `normalize` (US) to `normalise` (UK) to have the same spelling as the other files in aeon (both were sometime used in the same file).

MatthewMiddlehurst commented:
Will try get through it this week, but one comment before I dive in:

> Uncomment tests that were commented because of https://github.com/aeon-toolkit/aeon/issues/2071, and add some test exclusion due to issue with the estimator checking function giving wrong input shape for predict function of similarity search (3D fit, 2D predict.

It would be better to add to the testing data file rather than exclude? I can give some help if you need, but really all you need to do is add your data to the big dictionary then edit the functions below to make sure the correct key(s) is given to similarity searchers

MatthewMiddlehurst commented on aeon/testing/testing_config.py:
```diff
@@ -78,6 +73,21 @@
     "MultiRocketRegressor": ["check_non_state_changing_method"],
     "RSTSF": ["check_non_state_changing_method"],
+    # Due to numba random issue between OS
+    "RDSTRegressor": ["check_regressor_against_expected_results"],
```
Is this still an issue even with MacOS excluded?

baraline replied:
Yes I had failure on the CI with windows and ubuntu, MacOS gave correct results somehow if I recall correctly, problem is OS related. We could record results for different OS versions if we want to test it anyway while we try to fix it.

MatthewMiddlehurst replied:
We don't run the test on MacOS anymore since we moved over to the new runner, so if you are generating the results on a Mac it probably wont be correct.

baraline replied:
OK, I'll try to re-generate results on ubuntu and see if we get something consistent then.

baraline commented:
I'll try to have a look at to include back the ones I excluded later this week then, I'll send you a message if I'm stuck.

baraline commented:
@MatthewMiddlehurst added data generation and test for the similarity search estimators, removed the non-deterministic tags from RDST regressor and classifier and added some check against expected results for the classifier.
Only issue arise for RDST Regressor on covid3month specifically. See #2290 for details.

MatthewMiddlehurst commented on aeon/testing/utils/estimator_checks.py:
```diff
@@ -18,15 +19,29 @@ def _run_estimator_method(estimator, method_name, datatype, split):
+        else:
+            value = method()
+
+        # Similarity search return tuple as (distances, indexes)
+        if isinstance(estimator, BaseSimilaritySearch):
+            if isinstance(value, tuple):
+                return value[0]
+            else:
+                return value
```
does it fail without this? not sure what it would be checking

baraline replied:
It would fail without this as similarity search estimators returns a tuple of two array (match_distances, match_indexes), which is not standard as other methods just return one array (with predict/transform)

MatthewMiddlehurst replied:
I'm just wondering where in the general testing it would check for this really. It also covers transformers which have all sorts of outputs

MatthewMiddlehurst reviewed (commented):
Should be fine to let this through after the docs are merged. I will add the testing labels just in-case.

MatthewMiddlehurst reviewed (commented):
will leave RDST for now

baraline commented:
Deal, i'll take a look at the conflicts and we should be good to go then.

MatthewMiddlehurst commented:
If you are up for it, i think we are giving in the American spelling also 🙂

baraline commented:
@MatthewMiddlehurst answred comment and fixed tests that were causing failures (I implemented these test wrong). Concerning language, I would prefer to delay it to another PR, this one is big enough.

MatthewMiddlehurst reviewed (approved):
whats the worst its going to do, break everything?

baraline commented:
Well it seems that it run fines on main, sooooo all good I guess ?
